vm/stdlib/modules/Account.move

```diff
@@ -117,6 +125,46 @@ module Account {
 
     const DUMMY_AUTH_KEY:vector<u8> = x"0000000000000000000000000000000000000000000000000000000000000000";
 
+    // cannot be dummy key
+    const AUTH_KEY_PLACEHOLDER:vector<u8> = x"1000000000000000000000000000000000000000000000000000000000000000";
+
+    /// A one-way action, once SignerCapability is removed from signer, the address cannot send txns anymore.
+    public fun remove_signer_capability(s: &signer): SignerCapability
+    acquires Account {
+        let signer_addr = Signer::address_of(s);
+        // check signer not delegated.
+        let already_delegated = exists<SignerDelegated>(signer_addr);
+        assert(!already_delegated, 401);
+
+        // set to account auth key to noop.
+            {
+                let key_rotation_capability = extract_key_rotation_capability(s);
+                rotate_authentication_key_with_capability(&key_rotation_capability, AUTH_KEY_PLACEHOLDER);
+                restore_key_rotation_capability(key_rotation_capability);
+            };
+        move_to(s, SignerDelegated {});
+        let signer_cap = SignerCapability {addr: signer_addr };
+        signer_cap
+    }
```

Comment:
Couldn't `s` rotate its authentication key back to a known value immediately after calling this function? E.g.

```
fn main(s: signer) {
  let old_auth_key = Account::authentication_key(Signer::address_of(&s));
  let signer_cap = Account::remove_signer_capablity(&s); let key_rotation_capability = 
  let cap = extract_key_rotation_capability(s);
  rotate_authentication_key_with_capability(&cap, old_auth_key);
  ... // now s has both a SignerCapability and the ability to send txes from its account
}
```

Reply:
if we remove_signer_capablity and delegate signer to a Contract Module, it means we only want the Module can do something as the signer, so we need to do:

1.  Set the account's auth key as empty bytes, let the private key invalid.
2. Destroy the key_rotation_capability.
3. remove_signer_capablity and delegate signer cap to the Module.

Now the account becomes a Contract Account like Ethereum.

Reply:
> Couldn't `s` rotate its authentication key back to a known value immediately after calling this function? E.g.
> 
> ```
> fn main(s: signer) {
>   let old_auth_key = Account::authentication_key(Signer::address_of(&s));
>   let signer_cap = Account::remove_signer_capablity(&s);
>   let cap = extract_key_rotation_capability(s);
>   rotate_authentication_key_with_capability(&cap, old_auth_key);
>   ... // now s has both a SignerCapability and the ability to send txes from its account
> }
> ```

good catch!
Like jolestar said, more security check should be considered if we really want to do this.

Reply:
The main point of it is that let modules hold the signer cap of an address, which invalidate the signing cap of the EOA.
What concerns us is that it can has unsee influences while more features are added to Move language, and we don't know if it can fit the programming model of Move.

Reply:
The main point of it is that let modules hold the signer cap of an address, which invalidate the signing cap of the EOA.
What concerns us is that it can has unsee influences while more features are added to Move language, and we don't if it can fit the programming model of Move.

Reply:
> The main point of it is that let modules hold the signer cap of an address, which invalidate the signing cap of the EOA.

To understand the intent better, let me ask a question: would simply rotating the auth key of the EOA to `AUTH_KEY_PLACEHOLDER` accomplish the same thing? 

Or do you also want the module to have the `signer` so it can:
- take actions on behalf of the EOA that require a `signer` (e.g., send funds)
- use `move_to`
?

Reply:
We want a completely programable account, the smart contract module can actor like an account, do things like account, etch:

1. take actions on behalf of the EOA that require a signer.
2. use move_to.